projects/bootloader/src/bootloader_crc32.c

```diff
@@ -0,0 +1,34 @@
+#include "bootloader_crc32.h"
+
+#include "stdio.h"
+
+#include "bootloader_mcu.h"
+#include "crc32.h"
+#include "flash.h"
+
+// calculates the crc32 for the full address, with 2048 byte increments
+// at a time, returning the final crc32 code
+uint32_t calculated_application_crc32(uintptr_t address, size_t size) {
+  // code array for holding crc32 codes per 2048 bytes
+  // size of array is 54 since
+  // BOOTLOADER_APPLICATION_SIZE / 2048 = 54
+  uint32_t crc32_codes[54];
+  uint8_t crc32_code_number = 0;
+  size_t curr_size = 0;  // memory size that has been read
+  uint8_t buffer[2048];  // buffer for holding 2048 bytes of flash
+  uint32_t crc_temp;
+
+  while (curr_size <= size) {
+    // read from flash
+    flash_read(address + curr_size, sizeof(buffer), (uint8_t *)&buffer, sizeof(buffer));
+    // claculate crc32
+    crc_temp = crc32_arr((uint8_t *)&buffer, sizeof(buffer));
+
+    crc32_codes[crc32_code_number] = crc_temp;
+    crc32_code_number++;
+    curr_size += sizeof(buffer);
+  }
+
+  // calculate the final crc32 and return it
+  return crc32_arr((uint8_t *)&crc32_codes, sizeof(crc32_codes));
```

Comment:
Similar thing here: use `(uint8_t *)crc32_codes` (no `&`) instead because `crc32_codes` is implicitly converted to a `uint32_t *`.

Also, the behaviour here will depend on the [endianness](https://en.wikipedia.org/wiki/Endianness) of the system: since you're treating `uint32_t`s as arrays of four `uint8_t`s, you'll get different results on a little-endian system, where a uint32_t like `0x12345678` is laid out like `78 56 34 12` in memory, than on a big-endian system, where `0x12345678` would be laid out like `12 34 56 78` in memory. This hopefully shouldn't be too much of an issue since most systems are little-endian, but to make it robust, I'd recommend creating a second array of `uint8_t`s and manually copying each byte of each uint32_t in `crc32_codes` to the uint8_t array (using bitwise operations to extract each byte). Since you're manually dealing with the bytes in this case, it should be the same regardless of endianness.